README

		*************************************
		  The Tk::IDElayout Widget
		*************************************

INTRODUCTION

Tk::IDElayout is a widget for managing the layout of Tk frames
(and other widgets) like an IDE (Integrated Development Environment)
similar to Ecliplse or Microsoft Visual Studio.

Features:

See "Screenshots.pdf" in the source distribution for some screenshots
demonstrating some of these features.

* Layout and management of subwidgets/frames similar to an IDE.
	
* Resizable panes: Separate frames/widgets in the top-level window can be resized by dragging
the separator border between the frames.
	
* Support for Tabbed-Frames (using of subclass of L<Tk::DynaTabFrame>, where each tab
can be dragged/dropped; to another tabframe in the IDE, or to a new separate window,
or to "edge" ares in the mainwindow to create new tabframes.


BUILDING AND INSTALLING THE WIDGET

1. Uncompress and unpack the distribution

   ON UNIX:
	gzip -d Tk-TableMatrix<version>.tar.gz
	tar -xf Tk-TableMatrix<version>.tar

   ON WINDOWS:
	use something like WinZip to unpack the archive.

    
   This will create a subdirectory TableMatrix<version> with all the files in it.

2. Build/Test

  Note: A c-compiler is required, due to the CaptureRelease extension, which enables windows
  to be released from the IDE to separate windows and captured back to the IDE.

    perl Makefile.PL
    make 
    make test
    
 The test case t/simpleIDElayout2.t serves a simple demo. Run it using
    perl -Mblib t/simpleIDElayout2.t
   
3. Install
    make install

4. Read the documentation

   Read it using the standard 'perldoc Tk::IDElayout' command.

           
           LICENSE

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Summary of Changes to Tk::IDElayout

0.37_02  2020-12-16

* Prevent UNKNOWN reports from CPAN testers:
  - Incorporate change from Perl/Tk's myConfig:
    exit gracefully if libX11 not found
  - specify dependencies for configure phase
    (i.e. needed when running Makefile.PL, myConfig)

0.37_01  2020-12-13

 * Makefile.PL: Use non-zero exit code when myConfig fails
   (trying to debug UNKNOWN reports from CPAN testers)

0.37  2020-12-12

 * Use Graph 0.50 syntax (mandatory as of Graph 0.9712)

0.36_01  2020-12-11

 * Specify GitHub issues as bugtracker

0.36  2020-11-14

 * Fix for Darwin/macOS build:
   Avoid error caused by trying to link to Tk library in Mach-O bundle
   ("ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file")

0.35  2020-03-09

 * Fix POD error (missing `=back` before `=head1`) in IDEtabFrame.pm
   (contribution from Mohammad S Anwar)
 * Makefile.PL: Don't index unauthorized packages Tk and Tk::CaptureRelease
   to avoid CPAN permission issue

0.34  2018-08-02

 * Improved error message if widgets are missing containers needed for dropping
   (contribution from Alexander Becker)
 * Include GNU GPLv1 (not GPLv2) in LICENSE file to match Perl 5 license
 * Update CPAN META info: add GitHub repository, license
 
0.33  2014-04-26

 * Fixed intermittent bug where in the populate method, the frameStructure method would get called before the -widgets
   stucture would get populated, causing errors.

0.32  2008-09-18

 * Fixed bug where the status line would vanish if the whole window was expanded and then shrunk.
 * Updated Makefile.PL so CPAN (hopefully) will understand that Tk is required for this module.
     (Since Tk was required for just running the Makefile.PL, this was more complicated than just
      adding to PREREQ_PM.)

0.31  2007-09-20

 * Updated to work with Graph packages greater than version .3
 * Fixed memory leak where widgets where not being destroyed after the tabs that contained
   them were closed.

0.3  2007-09-16

 * First Public Release.




Makefile.PL
use 5.6.0;
use Cwd;
use Config;
no  lib '.';
use ExtUtils::MakeMaker;
BEGIN {

    warn("### Note: 'Building Outside of Tk itself' Messages are Expected Here ####\n");

    $IsWin32 = ($^O eq 'MSWin32' || $Config{'ccflags'} =~ /-D_?WIN32_?/);

    $VERSION = '0.37_02';
    $win_arch = shift @ARGV if @ARGV and $ARGV[0] =~ /^(open32|pm|x|MSWin32)$/;
    require('fix_4_os2.pl'), OS2_massage() if $^O eq 'os2';

    # Currently 'x', 'pm', 'open32', 'MSWin32'
    $win_arch = ($IsWin32) ? 'MSWin32' : 'x' if not defined $win_arch;

    require './myConfig';

  # use lib ($Tk::MMutil::dir=getcwd);

} # /BEGIN

use Tk::MMutil;

# Path to the installed shared Tk library (needed on unix)
my $TkLibPath = ' ';

if ($IsWin32) {
    *MY::makeaperl = \&makeWin32perl;
    if ($Config{cc} =~ /^gcc/i) {
        @libs = ('');
    }
} else {
    # We need to link in the Tk Library on Unix (unless it is a Darwin/macOS
    # Mach-O bundle not meant for linking) to get TkpWmSetState and other symbols
    #    (These are exported in the TkintplatdeclsVptr in windows, but not on unix)
    $TkLibPath = Tk::MMutil::findINC('auto/Tk/Tk.'.$Config{dlext});
    if ($^O eq 'darwin') {
        # Check for MH_BUNDLE header fields (based on `struct mach_header` and
        # `struct mach_header_64` from <mach-o/loader.h>):
        #    - magic    (uint32_t at bytes 0 to 4)
        #    - filetype (uint32_t at bytes 12 to 15)
        my $MH_MAGIC    = 0xfeedface;
        my $MH_CIGAM    = 0xcefaedfe;
        my $MH_MAGIC_64 = 0xfeedfacf;
        my $MH_CIGAM_64 = 0xcffaedfe;
        my $MH_BUNDLE   = 0x8;
        open my $TkLibFH, '<:raw', $TkLibPath;
        my $header_size = read $TkLibFH, my $header, 16;
        die "Couldn't read header of $TkLibPath" unless $header_size == 16;
        close $TkLibFH;
        my ($magic, $filetype) = unpack('L x8 L', $header);
        #printf("Got magic == 0x\%x, filetype == 0x\%x\n", $magic, $filetype);
        $TkLibPath = ' ' if (
            ($magic == $MH_MAGIC or
             $magic == $MH_CIGAM or
             $magic == $MH_MAGIC_64 or
             $magic == $MH_CIGAM_64)
            and ($filetype == $MH_BUNDLE));
    }

    my $plibs = $Config{'libs'};
    my $libs  = "$xlib -lX11";   # Used to have -lpt in here as well.
    my @try   = qw(-lsocket -lnsl -lm);
    push(@try,'-lc') if $^O eq 'svr4';
    my $lib;
    # perl needs sockets and math library too
    # so only include these if they made it through perl's Configure
    foreach $lib (@try) {
        $libs .= " $lib" if ($plibs =~ /$lib\b/);
    }
    @libs = ("$libs");
}

my $dir = Tk::MMutil::find_subdir();
delete $dir->{'pTk'};

# Find path to existing pTk include files
my $ptkPath = Tk::MMutil::findINC('Tk/pTk/Lang.h');
$ptkPath =~ s/\/Lang\.h$//g;



Tk::MMutil::TkExtMakefile(
    'VERSION'  => $VERSION,
    'XS_VERSION' => $VERSION,
    'EXE_FILES' => [],
    'NAME'     => 'Tk::CaptureRelease',
    'LICENSE'  => 'perl_5',
    'META_MERGE' => {
        'meta-spec' => { version => 2 },
        no_index => {
            'package' => [ 'Tk', 'Tk::CaptureRelease' ],
        },
        resources => {
            repository => {
                type => 'git',
                url => 'https://github.com/asb-capfan/Tk-IDElayout.git',
                web => 'https://github.com/asb-capfan/Tk-IDElayout',
            },
            bugtracker => {
                web => 'https://github.com/asb-capfan/Tk-IDElayout/issues',
            },
        },
    },
    'DIR'      => [reverse(sort(keys %$dir))],
    'DISTNAME' => "Tk-IDElayout",
    'DEFINE'   => $mydefine,
    'INC'    => "-I$ptkPath",
    'LIBS'    => \@libs,
    'OBJECT'  => '$(O_FILES)'. " $TkLibPath", # Tk Lib path will be the Tk library on Unix, blank for win32
    'CONFIGURE_REQUIRES' => {
        'ExtUtils::MakeMaker' => '6.52',
        'Tk::MMtry' => 0,
        'Tk::MMutil' => 0,
    },
    'PREREQ_PM' => {
        'Tk' => '804.027',
        'Tk::DynaTabFrame' => '0.23',
        'Graph' => '0.50',
    },
    @macro,
);


sub MY::post_initialize {
    my ($self) = @_;
    my ($ret) = '';
    my %files = ();
    my $dir = $self->catdir('$(INST_ARCHLIBDIR)','Tk');
    my $name;
    foreach $name (grep /(%|\.q4|\.bck|\.old)$/,keys %{$self->{PM}}) {
        delete $self->{PM}->{$name};
    }
    # delete $self->{PM}->{'Tk/Config.pm'};
    # $self->{PM}->{'Tk/Config.pm'} = $self->catfile($dir,'Config.pm');
    # $files{'typemap'} = 1;
    foreach $name ($self->lsdir(".")) {
        next if ($name =~ /^\./);
        next unless (-f $name);
        $files{$name} = 1 if ($name =~ /\.[tm]$/);
        $files{$name} = 1 if ($name =~ /\.def$/);
    }
    foreach $name (sort(@{$self->{H}},keys %files)) {
        $self->{PM}->{$name} = $self->catfile($dir,$name);
    }
    $ret;
}

sub needs_Test {
    my $file = shift;
    local $_;
    unless (open(TFILE,"$file")) {
        warn "Cannot open $file:$!";
        return 1;
    }
    my $code = 0;
    while (<TFILE>) {
        last if ($code = /^\s*(use|require)\s+Test\b/);
    }
    close(TFILE);
    warn "Skipping test $file needs 'Test.pm':$_" if $code;
    return $code;
}

sub MY::test {
    my ($self,%attrib) = @_;
    my @tests = sort glob($self->catfile('t','*.t'));
    eval { require Test };
    if ($@) {
        @tests = grep(!needs_Test($_),@tests);
    }
    $attrib{'TESTS'} = join(' ',@tests);

    # Temporarily remove sub-dirs from $self as we 'know'
    # there are no tests down there
    my $dir  = delete $self->{'DIR'};
    my $str  = $self->MM::test(%attrib);
    # Put sub-dirs back
    $self->{'DIR'} = $dir;
    return $str;
}

sub MY::postamble {
    '


Makefile     : myConfig

perlmain.c : config Makefile

basic      : $(INST_DYNAMIC) pm_to_blib

MANIFEST : MANIFEST.SKIP $(FIRST_MAKEFILE)
	$(MAKE) manifest

';
}
